lib/remap-parser/src/parser.lalrpop

```diff
@@ -0,0 +1,501 @@
+pub Test: Test = {
+    // root
+    "_r" <Expr> => Test::Expr(<>),
+
+    // expressions
+    "_e" <Literal> => Test::Literal(<>),
+    "_e" <Container> => Test::Container(<>),
+
+    // arithmetic (math)
+    "_m" <ArithmeticExpr> => Test::Arithmetic(<>),
+
+    // atoms
+    "_a" <String> => Test::String(<>),
+    "_a" <Integer> => Test::Integer(<>),
+    "_a" <Float> => Test::Float(<>),
+    "_a" <Boolean> => Test::Boolean(<>),
+    "_a" <Null> => Test::Null(()),
+    "_a" <Regex> => Test::Regex(<>),
+
+    // collections
+    "_c" <Block> => Test::Block(<>),
+    "_c" <Array> => Test::Array(<>),
+    "_c" <Object> => Test::Object(<>),
+
+    // other
+    "_as" <Assignment> => Test::Assignment(<>),
+    "_fn" <FunctionCall> => Test::FunctionCall(<>),
+    "_q" <Query> => Test::Query(<>),
+};
+
+// -----------------------------------------------------------------------------
+// root expressions
+// -----------------------------------------------------------------------------
+
+RootExprs: Vec<Node<RootExpr>> = {
+    RootExpr => vec![<>],
+    (<RootExpr> EndOfExpression)*,
+};
+
+RootExpr: Node<RootExpr> = {
+    Expr => Node::new(<>.span(), RootExpr::Expr(<>)),
+
+    // Root expressions are allowed to fail. The parser will continue with the
+    // next expression in the program.
+    Sp<!> => {
+        let Node { span, node } = <>;
+        let ErrorRecovery { error: source, dropped_tokens } = node;
+
+        let reserved_keyword = dropped_tokens
+            .first()
+            .cloned()
+            .into_iter()
+            .find_map(|(start, t, end)| match t {
+                Token::ReservedIdentifier(s) => Some((start, s.to_owned(), end)),
+                _ => None,
+            });
+
+        if let Some((start, keyword, end)) = reserved_keyword {
+            let error = Error::ReservedKeyword {
+                start,
+                keyword: keyword,
+                end,
+            };
+
+            return Node::new(span, RootExpr::Error(error));
+        }
+
+        let source = source
+            .map_token(|t| t.map(|s| s.to_owned()))
+            .map_error(|e| e.to_string());
+
+        let dropped_tokens = dropped_tokens
+            .into_iter()
+            .map(|(l, t, r)| (l, t.map(|s| s.to_owned()), r))
+            .collect();
+
+        let error = Error::ParseError {
+            source,
+            dropped_tokens,
+        };
+
+        Node::new(span, RootExpr::Error(error))
+    },
+};
+
+// -----------------------------------------------------------------------------
+// expressions
+// -----------------------------------------------------------------------------
+
+Exprs: Vec<Node<Expr>> = {
+    Expr => vec![<>],
+    <v:(<Expr> EndOfExpression)+> <e:(<Expr>)?> => match e {
+        None => v,
+        Some(e) => {
+            let mut v = v;
+            v.push(e);
+            v
+        }
+    }
+};
+
+
+Expr: Node<Expr> = {
+    Sp<IfStatement> => Node::new(<>.span(), Expr::IfStatement(<>)),
```

Comment:
`Sp<...>` is a _macro_, it expands to a bigger set of matching rules. See the end of this file for its implementation. The TL;DR is that it enriches the if statement rule with span details pointing to start/end locations in the source code. This is what allows our diagnostic messages to give detailed help messages.